core/chain_makers.go

```diff
@@ -239,7 +243,83 @@ func GenerateChain(config *params.ChainConfig, parent *types.Block, engine conse
 	return blocks, receipts
 }
 
-func makeHeader(chain consensus.ChainReader, parent *types.Block, state *state.StateDB, engine consensus.Engine) *types.Header {
+type knot struct {
+	block *types.Block
+}
+
+func GenerateKnot(config *params.ChainConfig, parent *types.Block, engine consensus.Engine, db ethdb.Database, n int, gen func(int, *BlockGen)) ([]*types.Block, []types.Receipts) {
+	if config == nil {
+		config = params.TestChainConfig
+	}
+	blocks, receipts := make(types.Blocks, n), make([]types.Receipts, n)
+	chainreader := &fakeChainReader{config: config}
+
+	resultLoop := func(i int, k *knot, statedb *state.StateDB, results chan *types.Block, stop chan struct{}, wg *sync.WaitGroup) error {
+		for {
+			select {
+			case block := <-results:
+				context, err := engine.GetDifficultyOrder(block.Header())
+				if err != nil {
+					log.Println("Block mined has an invalid order")
+				}
+
+				if context == 0 {
+					log.Println("PRIME:", block.Header().Location, block.Header().Number, block.Header().Hash())
+				}
+
+				k.block = block
```

Comment:
This saves every found block (even region & zone difficulties), but IIRC the knot requires 1 prime block from each location, right? So we should only save the block if it was a prime block.

Reply:
I believe the types.QuaiNetworkContext is 0 which enforces Prime blocks to be mined.

Reply:
Then why do you check if the block is prime a few lines above? L266

Reply:
It was a manager print that was copied over and I had the guard just to be sure we checked the context. My output looks like this:
```go
2022/08/26 11:28:37 PRIME: [1 1] [1 1 1] 0x4fbdd5d402446e7394c3b914af5bea84f66b3fd4944aade94c048af8413a5caf
2022/08/26 11:28:43 PRIME: [1 2] [2 2 1] 0x2684f68ee59e52899f426276b4899c5a0d8fa056216c00367dece7f72ddbfbb2
2022/08/26 11:29:08 PRIME: [1 3] [3 3 1] 0x145399974b54a19f88644ab1703fc6f5485b10d558fb3dc8754745fcd7209bb3
2022/08/26 11:29:16 PRIME: [2 1] [4 1 1] 0x9cfd18209b499753f8005bb4496d335c00a5ebba0aae9c7061f99e4908de320c
2022/08/26 11:29:18 PRIME: [2 2] [5 2 1] 0xc133649c4f0bf564fbe1304bec2c392b94eaa4bffbd6425c61973f6c79839a8e
2022/08/26 11:29:40 PRIME: [2 3] [6 3 1] 0x6a4b4f65e1361d54f4e3ca31ace2f9512be9997563cb7da77998bb1af363df0f
2022/08/26 11:30:18 PRIME: [3 1] [7 1 1] 0x6f4076f4cba634085337e27642545d947794106ec5f4502a2bb8910b22b18c7f
2022/08/26 11:30:30 PRIME: [3 2] [8 2 1] 0x5eb0610936254b0b0d69d471c3a0ca5ca0c13d68424f49e6a12155d8929bf502
2022/08/26 11:30:34 PRIME: [3 3] [9 3 1] 0xf779f268db7b98df1f156b37a0b0382eac7af0e5ee0ef76348ed44fbd4e39f46
```